File: src/lib/mechanics/leveling.ts
import { MAX_LEVEL } from "@/lib/constants";

export interface ClassLevel {
    classId: string;
    level: number;
}

export function calculateTotalLevel(classes: ClassLevel[]): number {
    return classes.reduce((sum, c) => sum + c.level, 0);
}

export function calculateProficiencyBonus(totalLevel: number): number {
    if (totalLevel <= 0) return 2;
    return Math.ceil(1 + totalLevel / 4);
}

export const XP_TABLE: Record<number, number> = {
    1: 0,
    2: 300,
    3: 900,
    4: 2700,
    5: 6500,
    6: 14000,
    7: 23000,
    8: 34000,
    9: 48000,
    10: 64000,
    11: 85000,
    12: 100000,
    13: 120000,
    14: 140000,
    15: 165000,
    16: 195000,
    17: 225000,
    18: 265000,
    19: 305000,
    20: 355000
};

export function getXpForLevel(level: number): number {
    return XP_TABLE[level] || 0;
}

/**
 * Check if character can level up ONCE (to the next level only).
 * Returns true only if XP meets the requirement for exactly the next level.
 */
export function canLevelUp(currentClasses: ClassLevel[], currentXp: number): boolean {
    const total = calculateTotalLevel(currentClasses);
    if (total >= MAX_LEVEL) return false;
    const requiredXp = getXpForLevel(total + 1);
    return currentXp >= requiredXp;
}

/**
 * Get how many levels the character COULD level up (for display purposes).
 * This shows potential levels but leveling happens one at a time.
 */
export function getPotentialLevels(currentClasses: ClassLevel[], currentXp: number): number {
    const currentTotal = calculateTotalLevel(currentClasses);
    let potential = 0;
    for (let lvl = currentTotal + 1; lvl <= MAX_LEVEL; lvl++) {
        if (currentXp >= getXpForLevel(lvl)) {
            potential++;
        } else {
            break;
        }
    }
    return potential;
}

File: shared/src/commonMain/kotlin/org/dndcharacter/shared/mechanics/Leveling.kt
package org.dndcharacter.shared.mechanics

import org.dndcharacter.shared.model.CharacterClass

object Leveling {
    const val MAX_LEVEL = 20

    val XP_TABLE = mapOf(
        1 to 0,
        2 to 300,
        3 to 900,
        4 to 2700,
        5 to 6500,
        6 to 14000,
        7 to 23000,
        8 to 34000,
        9 to 48000,
        10 to 64000,
        11 to 85000,
        12 to 100000,
        13 to 120000,
        14 to 140000,
        15 to 165000,
        16 to 195000,
        17 to 225000,
        18 to 265000,
        19 to 305000,
        20 to 355000
    )

    fun calculateTotalLevel(classes: Array<CharacterClass>): Int {
        return classes.sumOf { it.level }
    }

    fun calculateProficiencyBonus(totalLevel: Int): Int {
        if (totalLevel <= 0) return 2
        return kotlin.math.ceil(1 + totalLevel / 4.0).toInt()
    }

    fun getXpForLevel(level: Int): Int {
        return XP_TABLE[level] ?: 0
    }

    fun canLevelUp(currentClasses: Array<CharacterClass>, currentXp: Int): Boolean {
        val total = calculateTotalLevel(currentClasses)
        if (total >= MAX_LEVEL) return false
        val requiredXp = getXpForLevel(total + 1)
        return currentXp >= requiredXp
    }

    fun getPotentialLevels(currentClasses: Array<CharacterClass>, currentXp: Int): Int {
        val currentTotal = calculateTotalLevel(currentClasses)
        var potential = 0
        for (lvl in (currentTotal + 1)..MAX_LEVEL) {
            if (currentXp >= getXpForLevel(lvl)) {
                potential++
            } else {
                break
            }
        }
        return potential
    }
}
